pkg/dsl/lexer.go

```diff
@@ -0,0 +1,236 @@
+}
+
+type RegExpr kvPair
+
+func (e RegExpr) toTree() (*tree, error) {
+	r, err := regexp.Compile(e.value)
+	if err != nil {
+		return nil, fmt.Errorf("invalid regex filter: cannot compile regex [%w]", err)
+	}
+	return &tree{predicate: filters.Regex(e.key, r)}, nil
+}
+
+type NRegExpr kvPair
+
+func (e NRegExpr) toTree() (*tree, error) {
+	r, err := regexp.Compile(e.value)
+	if err != nil {
+		return nil, fmt.Errorf("invalid regex filter: cannot compile regex [%w]", err)
+	}
+	return &tree{predicate: filters.NotRegex(e.key, r)}, nil
+}
+
+type WithExpr struct {
+	key string
+}
+
+func (e WithExpr) toTree() (*tree, error) {
+	return &tree{predicate: filters.Presence(e.key)}, nil
+}
+
+type WithoutExpr struct {
+	key string
+}
+
+func (e WithoutExpr) toTree() (*tree, error) {
+	return &tree{predicate: filters.Absence(e.key)}, nil
+}
+
+type LogicalExpr struct {
+	left     Expression
+	operator string
+	right    Expression
+}
+
+func (le LogicalExpr) toTree() (*tree, error) {
+	left, err := le.left.toTree()
+	if err != nil {
+		return nil, err
+	}
+	right, err := le.right.toTree()
+	if err != nil {
+		return nil, err
+	}
+	return &tree{
+		logicalOp: le.operator,
+		children:  []*tree{left, right},
+	}, nil
+}
+
+type Lexer struct {
+	scanner.Scanner
+	errs   []error
+	result Expression
+}
+
+func (l *Lexer) Lex(lval *yySymType) int {
+	token := l.Scan()
+	if token == scanner.EOF {
+		return 0
+	}
+	tokenText := l.TokenText()
+	lval.value = tokenText
+
+	switch token {
+	case scanner.Int:
+		// Reading arbitrary number
+		res, err := strconv.ParseInt(tokenText, 10, 64)
+		if err != nil {
+			l.Error(err.Error())
+			return 0
+		}
+		lval.intValue = int(res)
+		return NUMBER
+	case scanner.Float:
+		l.Error("Float values are currently unsupported")
+		return 0
+	case scanner.String, scanner.RawString:
+		// Reading arbitrary double-quotes delimited string
+		var err error
+		lval.value, err = strconv.Unquote(tokenText)
+		if err != nil {
+			l.Error(err.Error())
+			return 0
+		}
+		return STRING
+	}
+
+	// Check if this is a syntaxToken
+
+	// Some characters are read as a token, such as "=", regardless of what follows
+	// To read "=~" as a token, we need to Peek next rune manually
+	tokenNext := tokenText + string(l.Peek())
+	if tok, ok := syntaxTokens[tokenNext]; ok {
+		l.Next()
+		return tok
+	}
+
+	if tok, ok := syntaxTokens[strings.ToLower(tokenText)]; ok {
+		return tok
+	}
+
+	// When none of the above returned, this must be a NetFlow field name
+	return NF_FIELD
+}
+
+func (l *Lexer) Error(msg string) {
+	l.errs = append(l.errs, fmt.Errorf("%s: %d:%d", msg, l.Line, l.Column))
+}
+
+func Parse(s string) (filters.Predicate, error) {
+	l := new(Lexer)
+	l.Init(strings.NewReader(s))
+	yyErrorVerbose = true
+	yyParse(l)
+	if len(l.errs) > 0 {
+		return nil, errors.Join(l.errs...)
+	}
+	t, err := l.result.toTree()
+	if err != nil {
+		return nil, err
+	}
+	return t.apply, nil
+}
```

Comment:
As this `Parse` function handle errors, I feel it would be usefull to parse the queries at operator level to ensure the data is correct before creating all the components.

WDYT ?

Reply:
yes it's done already :-)
https://github.com/netobserv/network-observability-operator/pull/1434/files#diff-9c4a6e3aff697b6f2534b57f05ef86381dd8c5d1dc37c60d586100ab72bb044aR283